src/agent/restart_handler/src/restart_handler_unix.cpp

```diff
@@ -0,0 +1,100 @@
+#include "restart_handler_unix.hpp"
+#include <fstream>
+#include <logger.hpp>
+
+#include <chrono>
+#include <thread>
+
+namespace restart_handler
+{
+
+    std::vector<char*> RestartHandler::cmd_line;
+
+    bool UsingSystemctl()
+    {
+        return (0 == std::system("which systemctl > /dev/null 2>&1") && nullptr != std::getenv("INVOCATION_ID"));
+    }
+
+    boost::asio::awaitable<module_command::CommandExecutionResult> RestartWithSystemd()
+    {
+        LogInfo("Systemctl restarting wazuh agent service.");
+        if (std::system("systemctl restart wazuh-agent") != 0)
+        {
+            co_return module_command::CommandExecutionResult {module_command::Status::IN_PROGRESS,
+                                                              "Systemctl restart execution"};
+        }
+        else
+        {
+            LogError("Failed using systemctl.");
+            co_return module_command::CommandExecutionResult {module_command::Status::FAILURE,
+                                                              "Systemctl restart failed"};
+        }
+    }
+
+    void StopAgent()
+    {
+        const int timeout = 30;
+        time_t start_time = time(nullptr); // Record the start time to track the timeout duration
+
+        pid_t pid = getppid();
+
+        // Shutdown Gracefully
+        kill(pid, SIGTERM);
+
+        while (true)
+        {
+            if (kill(pid, 0) != 0)
+            {
+                LogInfo("Agent gracefully stopped.");
+                break;
+            }
+
+            if (difftime(time(nullptr), start_time) > timeout)
+            {
+                LogError("Timeout reached! Forcing agent process termination.");
+                kill(pid, SIGKILL);
+            }
+
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        }
+    }
+
+    boost::asio::awaitable<module_command::CommandExecutionResult> RestartWithFork()
+    {
+        pid_t pid = fork();
+
+        if (pid < 0)
+        {
+            LogError("Fork failed.");
+        }
+        else if (pid == 0)
+        {
+            // Child process
+            StopAgent();
+
+            LogInfo("Starting wazuh agent in a new process.");
+
+            if (execve(RestartHandler::cmd_line[0], RestartHandler::cmd_line.data(), nullptr) == -1)
+            {
+                LogError("Failed to spawn new Wazuh agent process.");
```

Comment:
What if this happens, how the restart result is notified?

Reply:
In this case, it won't be possible since the agent is already stopped. What I can do is try to execute it three more times, maybe. What do you think?